lib/service.dart

```diff
@@ -192,6 +232,243 @@ class IsolateManager {
 
     return isolates.isEmpty ? null : isolates.first;
   }
+
+  void _initFlutterIsolate(List<IsolateRef> isolates) async {
+    for (IsolateRef ref in isolates) {
+      // Populate flutter isolate info.
+      if (_flutterIsolate == null) {
+        final Isolate isolate = await _service.getIsolate(ref.id);
+        if (isolate.extensionRPCs != null) {
+          for (String extensionName in isolate.extensionRPCs) {
+            if (extensionName.startsWith('ext.flutter.')) {
+              _setFlutterIsolate(ref);
+              break;
+            }
+          }
+        }
+      }
+      _serviceExtensionManager._addRegisteredExtensionRPCs(ref);
+    }
+  }
+
+  void _setFlutterIsolate(IsolateRef ref) {
+    if (_flutterIsolate == ref) {
+      // Isolate didn't change. Do nothing.
+      return;
+    }
+    _flutterIsolate = ref;
+    _flutterIsolateController.add(ref);
+    _flutterIsolateController.onListen =
+        () => _flutterIsolateController.add(_flutterIsolate);
+  }
+
+  StreamSubscription<IsolateRef> getCurrentFlutterIsolate(Function onData) {
+    return _flutterIsolateController.stream.listen(onData);
+  }
+}
+
+class ServiceExtensionManager {
+  VmService _service;
+  IsolateManager _isolateManager;
+
+  bool firstFrameEventReceived = false;
+
+  final Map<String, StreamController<bool>> _serviceExtensionController =
+      <String, StreamController<bool>>{};
+  final Map<String, StreamController<ServiceExtensionState>>
+      _serviceExtensionStateController =
+      <String, StreamController<ServiceExtensionState>>{};
+
+  /// All available service extensions.
+  final Set<String> _serviceExtensions = Set<String>();
+
+  /// All service extensions that are currently enabled.
+  final Map<String, ServiceExtensionState> _enabledServiceExtensions =
+      <String, ServiceExtensionState>{};
+
+  /// Temporarily stores service extensions that we need to add. We should not add
+  /// extensions until the first frame event has been received [firstFrameEventReceived].
+  final Set<String> _pendingServiceExtensions = Set<String>();
+
+  void _handleExtensionEvent(Event event) {
+    final String extensionKind = event.extensionKind;
+    if (event.kind == 'Extension' && extensionKind == 'Flutter.FirstFrame' ||
```

Comment:
add parens for clarity even though they aren't required.

Reply:
Done.